src/django_pg_migration_tools/operations.py

```diff
@@ -525,3 +583,256 @@ def describe(self) -> str:
+            schema_editor
+        )
+        if not self.allow_migrate_model(schema_editor.connection.alias, model):
+            return
+
+        table_name = model._meta.db_table
+        constraint_name = self._get_constraint_name(table_name, column_name)
+
+        # Just checking for NOT NULL isn't sufficient, we need to check that
+        # the constraint doesn't exist (VALID or NOT). Even a NOT VALID
+        # constraint would set the field to be non-nullable.
+        is_not_null = self._is_not_null(schema_editor, table_name, column_name)
+        constraint_exists = self._constraint_exists(schema_editor, constraint_name)
+        if is_not_null and (not constraint_exists):
+            return
+
+        if not constraint_exists:
+            self._alter_table_not_null_not_valid_constraint(
+                schema_editor, table_name, column_name, constraint_name
+            )
+            self._validate_constraint(schema_editor, table_name, constraint_name)
+            self._alter_table_not_null(schema_editor, table_name, column_name)
+            self._alter_table_drop_constraint(
+                schema_editor, table_name, constraint_name
+            )
+            return
+        elif self._is_constraint_valid(schema_editor, constraint_name):
+            self._alter_table_not_null(schema_editor, table_name, column_name)
+            self._alter_table_drop_constraint(
+                schema_editor, table_name, constraint_name
+            )
+            return
+        else:
+            # Constraint exists and is NOT VALID.
+            self._validate_constraint(schema_editor, table_name, constraint_name)
+            self._alter_table_not_null(schema_editor, table_name, column_name)
+            self._alter_table_drop_constraint(
+                schema_editor, table_name, constraint_name
+            )
+            return
+
+    def _alter_table_not_null_not_valid_constraint(
+        self,
+        schema_editor: base_schema.BaseDatabaseSchemaEditor,
+        table_name: str,
+        column_name: str,
+        constraint_name: str,
+    ) -> None:
+        cursor = schema_editor.connection.cursor()
+        cursor.execute(
+            psycopg_sql.SQL(
+                ConstraintQueries.ALTER_TABLE_CONSTRAINT_NOT_NULL_NOT_VALID
+            ).format(
+                table_name=psycopg_sql.Identifier(table_name),
+                column_name=psycopg_sql.Identifier(column_name),
+                constraint_name=psycopg_sql.Identifier(constraint_name),
+            )
+        )
+
+    def _is_not_null(
+        self,
+        schema_editor: base_schema.BaseDatabaseSchemaEditor,
+        table_name: str,
+        column_name: str,
+    ) -> bool:
+        cursor = schema_editor.connection.cursor()
+        cursor.execute(
+            NullabilityQueries.IS_COLUMN_NOT_NULL,
+            {"table_name": table_name, "column_name": column_name},
+        )
+        return bool(cursor.fetchone())
```

Comment:
Reviewer's [ 🍷 / 🧀 ]:

After executing that query which does a `SELECT 1 ... WHERE ...`, the `cursor.fetchone()` call returns either `None` or `(1,0)`. Even though those two possibilities are of different types, they correctly result in `False` or `True` when cast to `bool`.

(Reviewer is wary of this situation after creating a bug due to `bool((0,))` returning `True`.)